scripts/addr2line.py

```diff
@@ -83,6 +87,89 @@ def __call__(self, address):
         self._input.stdin.flush()
         return self._read_resolved_address()
 
+class KernelResolver:
+    """A resolver for kernel addresses which tries to read from /proc/kallsyms."""
+
+    LAST_SYMBOL_MAX_SIZE = 1024
+
+    def __init__(self):
+        syms : list[tuple[int, str]] = []
+        ksym_re = re.compile(r'(?P<addr>[0-9a-f]+) (?P<type>.+) (?P<name>\S+)')
+        warnings_left = 10
+
+        self.error = None
+
+        try:
+            f = open('/proc/kallsyms', 'r')
+        except OSError as e:
+            self.error = f'Cannot open /proc/kallsyms: {e}'
+            print(self.error)
+            return
+
+        try:
+            for line in f:
+                m = ksym_re.match(line)
+                if not m:
+                    if warnings_left > 0: # don't spam too much
+                        print(f'WARNING: /proc/kallsyms regex match failure: {line.strip()}', file=sys.stdout)
+                        warnings_left =- 1
+                else:
+                    syms.append((int(m.group('addr'), 16), m.group('name')))
+        finally:
+            f.close()
+
+        if not syms:
+            # make empty kallsyms (?) an error so we can assum len >= 1 below
+            self.error = 'kallsyms was empty'
+            print(self.error)
+            return
+
+        syms.sort()
+
+        if syms[-1][0] == 0:
+            # zero values for all symbols means that kptr_restrict blocked you
+            # from seeing the kernel symbol addresses
+            print('kallsyms is restricted, set /proc/sys/kernel/kptr_restrict to 0 to decode')
+            self.error = 'kallsyms is restricted'
+            return
+
+        # split because bisect can't take a key func before 3.10
+        self.sym_addrs : tuple[int]
+        self.sym_names : tuple[str]
+        self.sym_addrs, self.sym_names = zip(*syms) # type: ignore
+
+
+    def __call__(self, addrstr):
+        if self.error:
+            return addrstr + '\n'
+
+        sa = self.sym_addrs
+        sn = self.sym_names
+        slen = len(sa)
+        address = int(addrstr, 16)
+        # print('resolve address', address)
+        idx = bisect.bisect_right(sa, address) - 1
+        assert idx >= -1
+        assert idx < slen
```

Comment:
nit, could rewrite it like
```py
assert -1 <= idx < slen
```

IMHO, more concise this way.

Reply:
Thanks, changed. When I've written it the other way it's usually because I want to know which of the bounds failed, though ofc in python I can include this info in the assert message itself (again, mind poisoned by the spartan C assert).